src/main/scala/org/constellation/snapshotstreaming/storage/BaseLastIncrementalSnapshotStorage.scala

```diff
@@ -0,0 +1,98 @@
+package org.constellation.snapshotstreaming.storage
+
+import java.nio.file.NoSuchFileException
+import cats.effect.Async
+import cats.effect.std.Semaphore
+import cats.syntax.all._
+import cats.Applicative
+import cats.MonadThrow
+import org.tessellation.node.shared.domain.snapshot.Validator.isNextSnapshot
+import org.tessellation.node.shared.domain.snapshot.storage.LastSnapshotStorage
+import org.tessellation.schema._
+import org.tessellation.schema.height.Height
+import org.tessellation.security._
+import fs2.io.file._
+import fs2.Stream
+import fs2.text
+import io.circe.Codec
+import io.circe.parser.decode
+import io.circe.syntax._
+import org.tessellation.schema.snapshot.IncrementalSnapshot
+import org.tessellation.schema.snapshot.SnapshotInfo
+import org.tessellation.schema.snapshot.StateProof
+
+abstract class BaseLastIncrementalSnapshotStorage[F[_]: Async, P <: StateProof, S <: IncrementalSnapshot[P], SI <: SnapshotInfo[P]](semaphore: Semaphore[F], path: Path)(implicit codec: Codec[SnapshotWithState[S, SI]]) extends LastSnapshotStorage[F, S, SI] {
+
+  protected def validateStateProof(snapshot: Hashed[S], state: SI): F[Unit]
+
+  def set(snapshot: Hashed[S], state: SI): F[Unit] =
+    semaphore.permit.use { _ =>
+      validateStateProof(snapshot, state) >>
+        get.flatMap {
+          case Some(last) if isNextSnapshot(last, snapshot.signed.value) =>
+            Stream
+              .emit(SnapshotWithState(snapshot, state).asJson.spaces2)
+              .through(text.utf8.encode)
+              .through(Files[F].writeAll(path))
+              .compile
+              .drain
+          case Some(last) =>
+            MonadThrow[F].raiseError[Unit](
+              new Throwable(
+                s"Snapshot is not the next one! last: ${SnapshotReference
+                    .fromHashedSnapshot(last)}, lastHash: ${last.hash}, next: ${SnapshotReference
+                    .fromHashedSnapshot(snapshot)}, prevHash: ${snapshot.signed.value.lastSnapshotHash}"
+              )
+            )
+          case None =>
+            MonadThrow[F].raiseError[Unit](
+              new Throwable("Previous snapshot not found when setting next global snapshot!")
+            )
+        }
+    }
+
+  def setInitial(snapshot: Hashed[S], state: SI): F[Unit] =
+    semaphore.permit.use { _ =>
+      validateStateProof(snapshot, state) >>
+        Stream
+          .emit(SnapshotWithState(snapshot, state).asJson.spaces2)
+          .through(text.utf8.encode)
+          .through(Files[F].writeAll(path, Flags(Flag.Write, Flag.CreateNew)))
+          .compile
+          .drain
+    }
+
+  def get: F[Option[Hashed[S]]] =
+    getSnapshotWithState(_.snapshot)
+
+  def getCombined: F[Option[(Hashed[S], SI)]] =
+    getSnapshotWithState(sws => (sws.snapshot, sws.state))
+
+  def getCombinedStream: Stream[F, Option[(Hashed[S], SI)]] =
+    ???
```

Comment:
Yeah, it was like that before my refactorings. We reuse the `org.tessellation.node.shared.domain.snapshot.storage.LastSnapshotStorage` on both `snapshot-streaming` and `core`/`node`. In `core`/`node` it is used and implemented but here someone decided to make a shortcut by not implementing it as it is not used at all. That being said we can't remove it but we can provide a dummy implementation of `Stream.eval(getCombined)` :+1:

Reply:
Well, I think `Stream.eval(getCombined)` is an invalid implementation so it's better to provide nothing (`???`) rather than invalid one. If someone would try to use it, one will clearly see a not implemented error instead of figuring out what is wrong in the code. Otherwise, we need to provide the correct implementation that will stream elements and won't finish after the first emit like `Stream.eval(getCombined)`.